packages/@lwc/integration-karma/test/template/directive-if-elseif-else/index.spec.js

```diff
@@ -148,22 +149,112 @@ describe('lwc:if, lwc:elseif, lwc:else directives', () => {
+                // Text nodes are used as fragment delimiters
+                expect(assignedNodes[2].innerHTML).toBe('conditional slot content');
+            } else {
+                expect(assignedNodes.length).toBe(1);
+                expect(assignedNodes[0].innerHTML).toBe('static slot content');
+            }
+        }
+
+        it('should properly assign content for slots', () => {
+            const element = createElement('x-parent', { is: XparentWithSlot });
             document.body.appendChild(element);
 
-            const child = element.shadowRoot.querySelector('x-child-with-named-slots');
+            const child = element.shadowRoot.querySelector('x-child-with-slot');
 
             // When if condition is false, no slot content is provided by parent
-            const assignedNodes = child.shadowRoot.querySelector('slot').assignedNodes();
-            expect(assignedNodes).toHaveSize(0);
-
+            verifyExpectedSlotContent(child, false);
             element.condition = true;
             return Promise.resolve().then(() => {
+                // Slot content should be provided when condition is true
+                verifyExpectedSlotContent(child, true);
+            });
+        });
+
+        it('should properly rerender content for slots', () => {
+            const element = createElement('x-parent', { is: XparentWithSlot });
+            element.condition = true;
+            document.body.appendChild(element);
+
+            const child = element.shadowRoot.querySelector('x-child-with-slot');
+
+            verifyExpectedSlotContent(child, true);
+
+            element.condition = false;
+            return Promise.resolve()
+                .then(() => {
+                    verifyExpectedSlotContent(child, false);
+
+                    element.condition = true;
+                })
+                .then(() => {
+                    verifyExpectedSlotContent(child, true);
+                });
+        });
+
+        describe('named slots', () => {
+            /**
+             * Utility function to verify that slot content is assigned.
+             *
+             * @param {Element} child Child element to verify.
+             * @param {Boolean} condition Whether slot content is expected or not expected.
+             */
+            function verifyExpectedNamedSlotContent(child, condition) {
                 const assignedNodes = child.shadowRoot.querySelector('slot').assignedNodes();
-                expect(assignedNodes).toHaveSize(3); // VFragment has empty text nodes as delimiters
-                expect(assignedNodes[1].innerHTML).toBe('Named slot content from parent');
+                if (condition) {
+                    expect(assignedNodes.length).toBe(2);
+                    expect(assignedNodes[0].innerHTML).toBe('static slot content');
+                    expect(assignedNodes[1].innerHTML).toBe('conditional slot content');
```

Comment:
The two text nodes are going to the default slot because they do not have a `slot` attribute. I wonder if we should discard the text nodes when extracting the children from fragment in `collectSlots`. @jodarove @pmdartus What do you think?

Reply:
Note that if we do that, we end up with a difference behavior between synthetic and native shadow:

> I removed the splicing so that the synthetic shadow behavior matches the native shadow behavior. Originally I was hoping we could ignore the text nodes, but native includes them. Open to other suggestions or thoughts on whether we should and can hide the text nodes here in native:
> 
> **Slicing out the text nodes when extracting**: 
> no name - synthetic: `assignedNodes.length === 1`
> no name - native: `assignedNodes.length === 3` (includes the text nodes)
> named - synthetic: `assignedNodes.length === 1`
> named - native: `assignedNodes.length === 1`
> 
> **Current behavior with no slicing**:
> no name - synthetic: `assignedNodes.length === 3` (includes the text nodes)
> no name - native: `assignedNodes.length === 3` (includes the text nodes)
> named - synthetic: `assignedNodes.length === 1`
> named - native: `assignedNodes.length === 1`

You can see the behavior difference by running `yarn test:native`.

Reply:
mmm, this one is tricky.

the 2 text nodes delimiters are added for the diffing algo, but in this case, the allocation is the one doing the diffing and based on that the 2 delimiters could be discarded.

For native shadow, it is different, as the diffing is done to all the rendered children of the custom element. in this case, you do need the 2 delimiters.

I don't foresee any problems with the proposed solution, except but [this example](https://stackblitz.com/edit/salesforce-lwc-mfyveg?file=src%2Fmodules%2Fx%2Fcounter%2Fcounter.html,src%2Fmodules%2Fx%2Fapp%2Fapp.html,src%2Fmodules%2Fx%2Fapp%2Fapp.js&title=LWC%20playground):

```html
<template>
    <h2>Example 1, with if:true/if:false</h2>
    <x-counter>
        <template if:true={foo}>
            <template if:false={foo}>
                <span>hello</span>
            </template>
        </template>
    </x-counter>

    <h2>Example 2, with lwc:if</h2>
    <x-counter>
        <template lwc:if={foo}>
            <template if:false={foo}>
                <span>hello</span>
            </template>
        </template>
    </x-counter>
</template>
```

In the example, `foo` is `true`, and `x-counter` has a default slot with default content. Example 1 renders the default slot content, but example 2 renders the default slot empty.

what should be the correct behavior for Example 2?

Reply:
I don't know how to solve for it right now, but I think Example 2 should be rendering the default slot content. IMO the text delimiters shouldn't cause behavioral changes in the component.

>I don't foresee any problems with the proposed solution

Are you referring to discarding the text nodes?